share/p2p/shrex/server.go

```diff
@@ -0,0 +1,216 @@
+}
+
+func (srv *Server) handleNamespacedData(stream network.Stream) {
+	err := stream.SetReadDeadline(time.Now().Add(srv.readTimeout))
+	if err != nil {
+		log.Debugf("set read deadline: %s", err)
+	}
+
+	var req share_p2p_v1.GetSharesByNamespaceRequest
+	_, err = serde.Read(stream, &req)
+	if err != nil {
+		log.Errorw("read msg", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+
+	err = stream.CloseRead()
+	if err != nil {
+		log.Debugf("close read side of the stream: %s", err)
+	}
+
+	err = validateRequest(req)
+	if err != nil {
+		log.Errorw("invalid request", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), srv.serveTimeout)
+	defer cancel()
+
+	bs, dah, err := srv.store.CARBlockstore(ctx, req.RootHash)
+	if err != nil {
+		log.Errorw("get header for datahash", "err", err)
+		srv.respondInternal(stream)
+		return
+	}
+
+	// replace dah with testDAH within test scope
+	if srv.testDAH != nil {
+		dah = srv.testDAH
+	}
+
+	getter := eds.NewBlockGetter(bs)
+	blob, err := getBlobByNamespace(ctx, getter, dah, req.NamespaceId)
+	if err != nil {
+		log.Errorw("get shares", "err", err)
+		srv.respondInternal(stream)
+		return
+	}
+
+	resp := blobToResponse(blob)
+	srv.respond(stream, resp)
+}
+
+func getBlobByNamespace(
+	ctx context.Context,
+	getter blockservice.BlockGetter,
+	root *share.Root,
+	nID namespace.ID,
+) (*share.Blob, error) {
+	// select rows that contain shares with desired namespace.ID
+	selectedRows := make([]cid.Cid, 0)
+	for _, row := range root.RowsRoots {
+		if !nID.Less(nmt.MinNamespace(row, nID.Size())) && nID.LessOrEqual(nmt.MaxNamespace(row, nID.Size())) {
+			selectedRows = append(selectedRows, ipld.MustCidFromNamespacedSha256(row))
+		}
+	}
+	if len(selectedRows) == 0 {
+		return nil, nil
+	}
+
+	errGroup, ctx := errgroup.WithContext(ctx)
+	verifiedShares := make([]share.VerifiedShares, len(selectedRows))
+	for i, rowRoot := range selectedRows {
+		// shadow loop variables, to ensure correct values are captured
+		i, rowRoot := i, rowRoot
+		errGroup.Go(func() (err error) {
+			proof := new(ipld.Proof)
+			shares, err := share.GetSharesByNamespace(ctx, getter, rowRoot, nID, len(root.RowsRoots), proof)
+			verifiedShares[i].Shares = shares
+			verifiedShares[i].Proof = proof
+			return
+		})
+	}
+
+	if err := errGroup.Wait(); err != nil {
+		return nil, err
+	}
+	return &share.Blob{Rows: verifiedShares}, nil
+}
+
+// validateRequest checks correctness of the request
+func validateRequest(req share_p2p_v1.GetSharesByNamespaceRequest) error {
+	if len(req.NamespaceId) != ipld.NamespaceSize {
+		return fmt.Errorf("incorrect namespace id length: %v", len(req.NamespaceId))
+	}
+	if len(req.RootHash) != sha256.Size {
+		return fmt.Errorf("incorrect root hash length: %v", len(req.RootHash))
+	}
+
+	return nil
+}
+
+// respondInternal sends internal error response to client
+func (srv *Server) respondInternal(stream network.Stream) {
```

Comment:
rename to something that indicates that Internal means an error, "respondInternal" seems like it could just be a private response helper

Maybe `respondInternalErr` or `respondInternalError`